### PR TITLE
cmake: add and use new DT alias helpers

### PR DESCRIPTION
**Pull request by mbolivar-nordic**
Add convenience functions for dealing with devicetree aliases to extensions.cmake. Document them properly and mention them in the release notes. Document that they are expected to be added on an as-needed basis.

### CODE REVIEW

**Review by SebastianBoe — APPROVED** [review]

File: cmake/extensions.cmake [hunk 1]
```diff
@@ -1472,6 +1472,13 @@ function(target_sources_ifdef feature_toggle target scope item)
   endif()
 endfunction()
 
+function(target_sources_if_dt_alias alias target scope item)
```
Comment (tejlmand) [inline cmake/extensions.cmake:1475]: Could you please add some description of this function.

I'm curios on one thing.
This adds a `_if_dt_alias` which internally uses `dt_prop(... "/aliases" ...)` but where will this end then ?
Will there be a dedicated function for each path / property someone might ask for ?

Like:
`target_sources_if_dt_status_ok()` calling `dt_prop(... PATH <path> STATUS ...)

or other dt_prop dependent calls with other combinations

I think I understand the benefit, but what i'm trying to understand is how many places you expect this to be used, or if  this function should be a bit more generic so we won't need to add `n` dt functions for each combination of `dt_prop()`.

Maybe:
`target_sources_if_dt_prop(<path> <prop> <target> <scope> <item>)`
which in your case could be invoked as:
`target_sources_if_dt_prop("/aliases" <prop> <target> <scope> <item>)` (`<prop>` being your current `<alias>`)

**Comment by mbolivar-nordic** [comment]
I was firing this off quickly yesterday to get initial reviews. Thanks for those.

Since @tejlmand had questions about scope and ambition, I've spent more time trying to produce a more thorough PR. Hopefully it is acceptable now. Please revisit your reviews.

File: cmake/extensions.cmake (hunk 1 above)
Comment (mbolivar-nordic) [reply cmake/extensions.cmake]: I have reworked this PR substantially to address your concerns. Please revisit.

File: cmake/extensions.cmake [hunk 2]
```diff
@@ -2940,6 +2943,104 @@ function(dt_chosen var)
...
+#
+# Example usage:
+#
+#  # Prints a fatal error and halts cmake unless a led0 devicetree
+#  # alias exists.
+#  message_if_no_dt_alias("led0" FATAL_ERROR "no LED was defined")
+#
+# <alias>     : Devicetree alias to check
+# ...         : Additional arguments are passed to message()
+function(message_if_no_dt_alias alias)
```
Comment (pdgendt) [inline cmake/extensions.cmake:3037]: This function is now used in the samples, which previously verified the `status = okay` property. Does `dt_alias` also take that property into account?

**Review by tejlmand — CHANGES_REQUESTED** [review]
I think `dt_alias()` should support `REQUIRED` flag instead, like:
```
dt_alias(<var> ALIAS <alias> REQUIRED)
```

Also, from the updates in this PR, then there is not really a need for `target_sources_if_dt_alias()` as all the places it is being used are not really having optional inclusion of sources based on devicetree alias, all places require the devicetree alias being present.

**Comment by mbolivar-nordic** [comment]
The latest rework follows some more discussion with @tejlmand offline.

- all changes to samples removed; this can't work with twister as-is and it's too much work to make this situation better
- we now treat aliases interchangeably with full paths, for consistency with the general DT practice as well as the python APIs
- message_if_dt_alias removed
- target_sources_if_dt_alias is now target_sources_if_dt_node, which takes a path that may be an alias

File: cmake/extensions.cmake [hunk 3]
```diff
@@ -2990,6 +2990,77 @@ function(dt_chosen var)
...
+#   };
+#
+# Example usage:
+#
+#   dt_path_internal(ret "/foo/bar")     # sets ret to "/foo/bar"
+#   dt_path_internal(ret "my-alias")     # sets ret to "/foo/bar"
+#   dt_path_internal(ret "my-alias/baz") # sets ret to "/foo/bar/baz"
+#   dt_path_internal(ret "/blub")        # ret is undefined
+function(dt_path_internal var path)
+  string(FIND "${path}" "/" slash_index)
```
Comment (tejlmand) [inline cmake/extensions.cmake:3066]: an alternative would be to use a `REGEX MATCH` like this:
```
string(REGEX MATCH "^([/])?([^/]+)(/.*)*" match_result "${PATH}")
```

For example that will give:
```
path: /foo                                                                                                                                                                                                                                   
CMAKE_MATCH_1: /                                                                                                                                                                                                                             
CMAKE_MATCH_2: foo                                                                                                                                                                                                                           
CMAKE_MATCH_3:                                                                                                                                                                                                                               

path: /foo/bar                                                                                                                                                                                                                               
CMAKE_MATCH_1: /                                                                                                                                                                                                                             
CMAKE_MATCH_2: foo                                                                                                                                                                                                                           
CMAKE_MATCH_3: /bar                                                                                                                                                                                                                          

path: my-alias                                                                                                                                                                                                                               
CMAKE_MATCH_1:                                                                                                                                                                                                                               
CMAKE_MATCH_2: my-alias                                                                                                                                                                                                                      
CMAKE_MATCH_3:                                                                                                                                                                                                                               

path: my-alias/baz                                                                                                                                                                                                                           
CMAKE_MATCH_1:                                                                                                                                                                                                                               
CMAKE_MATCH_2: my-alias                                                                                                                                                                                                                      
CMAKE_MATCH_3: /baz      
```

So you can use `CMAKE_MATCH_1` to identify if `path` starts with `/`.
`CMAKE_MATCH_2` as `alias_name` and if `CMAKE_MATCH_3` is defined, then it's an alias path.

Doing so could cleanup some logic below I believe.

File: doc/releases/release-notes-3.0.rst [hunk 4]
```diff
@@ -287,6 +287,21 @@ USB
 Build and Infrastructure
 ************************
 
+* Build system
```
Comment (tejlmand) [inline doc/releases/release-notes-3.0.rst:290]: Nothing wrong on this line, but commit message still mentions: `message_if_no_dt_alias()`.

> Mention the dt_alias(), target_sources_if_dt_alias(), and
message_if_no_dt_alias() helpers.

**Review by tejlmand — COMMENTED** [review]
Generally good looking :+1: 

Some minor nits, and an alternative regex proposal.

File: cmake/extensions.cmake [hunk 5]
```diff
@@ -2990,6 +2990,77 @@ function(dt_chosen var)
...
+    # Otherwise, try to expand a leading alias.
+    string(SUBSTRING "${path}" 0 "${slash_index}" alias_name)
+    dt_alias(alias_path PROPERTY "${alias_name}")
+
+    # If there is a leading alias, append the rest of the string
+    # onto it and see if that's an existing node.
+    if (DEFINED alias_path)
+      set(rest)
+      if (NOT "${slash_index}" EQUAL -1)
+	string(SUBSTRING "${path}" "${slash_index}" -1 rest)
```
[the remark this answers is not in the document]
Comment (mbolivar-nordic) [reply cmake/extensions.cmake:3086]: Whoops, looks like I am using tabs here, will fix.

File: cmake/extensions.cmake (hunk 3 above)
Comment (mbolivar-nordic) [reply cmake/extensions.cmake:3066]: > an alternative would be to use a REGEX MATCH like this:

That regular expression doesn't work for the root node, `/`. I would prefer to leave this as-is.

**Comment by mbolivar-nordic** [comment]
@tejlmand thanks for review, comments addressed

File: cmake/extensions.cmake (hunk 3 above)
Comment (tejlmand) [reply cmake/extensions.cmake:3066]: > That regular expression doesn't work for the root node, `/`.

I wasn't aware the root node was a use-case also.
In that case the regex should look as
```
string(REGEX MATCH "^([/])?([^/]*)(/.*)*" match_result "/") 
```

> I would prefer to leave this as-is.

I'm fine with that as well.

Comment (mbolivar-nordic) [reply cmake/extensions.cmake:3066]: >> I would prefer to leave this as-is.
>
> I'm fine with that as well.

OK, that was your last unresolved comment, then, I think. OK to approve?

**Comment by tejlmand** [comment]
> @tejlmand thanks for review, comments addressed

except your last push didn't fix the indentation.
https://github.com/zephyrproject-rtos/zephyr/compare/db6a9c39e83487594d8a307e2eabc5e720150634..28f3cb9d85d62b668b6b82b9230b791b6483b59f

**Comment by mbolivar-nordic** [comment]
> except your last push didn't fix the indentation.

Darn, sorry, I threw away the fix when I realized the regex wouldn't handle the root node. I got it this time, thanks for catching that.

**Review by tejlmand — APPROVED** [review]

**Pull request merged**